File: ios/SpentyAI/SpentyAI/Core/Models/Record.swift
import Foundation

struct Record: Codable, Identifiable {
    let id: String
    var subject: String?
    var fromEmail: String?
    var archivedAt: Date?
    var date: String?
    var source: String?
    var transactionAmount: Double?
    var transactionType: String?
    var hasAttachments: Bool?
    var attachments: [RecordAttachmentRef]?

    /// Convenience accessors used by the UI
    var sender: String? { fromEmail }
    var receivedDate: Date? { archivedAt }
    var amount: Double? { transactionAmount }
    var attachmentCount: Int? { attachments?.count }

    enum CodingKeys: String, CodingKey {
        case id = "archiveId"
        case subject
        case fromEmail
        case archivedAt
        case date
        case source
        case transactionAmount
        case transactionType
        case hasAttachments
        case attachments
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id)
        subject = try? c.decode(String.self, forKey: .subject)
        fromEmail = try? c.decode(String.self, forKey: .fromEmail)
        archivedAt = try? c.decode(Date.self, forKey: .archivedAt)
        date = try? c.decode(String.self, forKey: .date)
        source = try? c.decode(String.self, forKey: .source)
        transactionAmount = try? c.decode(Double.self, forKey: .transactionAmount)
        transactionType = try? c.decode(String.self, forKey: .transactionType)
        hasAttachments = try? c.decode(Bool.self, forKey: .hasAttachments)
        attachments = try? c.decode([RecordAttachmentRef].self, forKey: .attachments)
    }
}

/// Lightweight attachment reference returned inside a Record list item
struct RecordAttachmentRef: Codable {
    let filename: String?
    let mimeType: String?
    let size: Int?
}

struct Receipt: Codable, Identifiable {
    let id: String
    var filename: String?
    var mimeType: String?
    var uploadedAt: Date?
    var parsedData: ReceiptParsedData?
    var transactionId: String?
    var fileSize: Int?

    enum CodingKeys: String, CodingKey {
        case id = "receiptId"
        case filename
        case mimeType
        case uploadedAt
        case parsedData
        case transactionId
        case fileSize
    }
}

struct ReceiptParsedData: Codable {
    var vendor: String?
    var amount: Double?
    var date: Date?
    var items: [ReceiptItem]?
    var tax: Double?
    var total: Double?
    var categoryName: String?
    var description: String?
    var paymentMethod: String?

    /// UI convenience – backend uses `vendor`, UI displays as merchant
    var merchant: String? { vendor }
}

struct ReceiptItem: Codable, Identifiable {
    var id: String { item ?? UUID().uuidString }
    var item: String?
    var amount: Double?
    var qty: Double?

    /// UI convenience accessors
    var description: String? { item }
    var quantity: Double? { qty }
}

File: android-native/app/src/main/java/com/spentyai/app/core/models/Record.kt
package com.spentyai.app.core.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Record(
    @SerialName("archiveId") val id: String = "",
    val subject: String? = null,
    val fromEmail: String? = null,
    val archivedAt: String? = null,
    val date: String? = null,
    val source: String? = null,
    val transactionAmount: Double? = null,
    val transactionType: String? = null,
    val hasAttachments: Boolean? = null,
    val attachments: List<RecordAttachmentRef>? = null
) {
    val sender: String? get() = fromEmail
    val amount: Double? get() = transactionAmount
    val attachmentCount: Int? get() = attachments?.size
}

@Serializable
data class RecordAttachmentRef(
    val filename: String? = null,
    val mimeType: String? = null,
    val size: Int? = null
)

@Serializable
data class Receipt(
    @SerialName("receiptId") val id: String = "",
    val filename: String? = null,
    val mimeType: String? = null,
    val uploadedAt: String? = null,
    val parsedData: ReceiptParsedData? = null,
    val transactionId: String? = null,
    val fileSize: Int? = null
)

@Serializable
data class ReceiptParsedData(
    val vendor: String? = null,
    val amount: Double? = null,
    val date: String? = null,
    val items: List<ReceiptItem>? = null,
    val tax: Double? = null,
    val total: Double? = null,
    val categoryName: String? = null,
    val description: String? = null,
    val paymentMethod: String? = null
) {
    val merchant: String? get() = vendor
}

@Serializable
data class ReceiptItem(
    val item: String? = null,
    val amount: Double? = null,
    val qty: Double? = null
) {
    val id: String get() = item ?: java.util.UUID.randomUUID().toString()
    val description: String? get() = item
    val quantity: Double? get() = qty
}

@Serializable
data class RecordListResponse(
    val records: List<Record> = emptyList(),
    val total: Int = 0
)

@Serializable
data class RecordSearchResponse(
    val items: List<Record> = emptyList(),
    val total: Int = 0
)

@Serializable
data class ReceiptListResponse(
    val receipts: List<Receipt> = emptyList(),
    val total: Int = 0
)

@Serializable
data class RecordPreviewResponse(
    @SerialName("archiveId") val id: String = "",
    val subject: String? = null,
    val fromEmail: String? = null,
    val archivedAt: String? = null,
    val source: String? = null,
    val body: String? = null,
    val bodyHtml: String? = null,
    val attachments: List<RecordAttachment>? = null,
    val transactionId: String? = null
) {
    val sender: String? get() = fromEmail
}

@Serializable
data class RecordAttachment(
    val index: Int = 0,
    val filename: String? = null,
    val mimeType: String? = null,
    val size: Int? = null
) {
    val id: String get() = "$index-${filename ?: "unknown"}"
}

@Serializable
data class ReceiptUploadResponse(
    @SerialName("receiptId") val id: String = "",
    val filename: String? = null,
    val fileSize: Int? = null,
    val mimeType: String? = null
)

@Serializable
data class ReceiptParseResponse(
    val receiptId: String = "",
    val parsedData: ReceiptParsedData? = null,
    val error: String? = null
)

@Serializable
data class ReceiptLinkBody(
    val transactionId: String
)

@Serializable
data class ReceiptLinkResponse(
    val message: String? = null,
    val receiptId: String? = null,
    val transactionId: String? = null
)

@Serializable
data class DownloadZipBody(
    val archiveIds: List<String>
)
